File: tests/wasm/test_volatility_adjusted_ma.js

import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import {
    loadTestData,
    assertArrayClose,
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN
} from './test_utils.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);


import * as wasm from '../../pkg/vector_ta.js';


const EXPECTED_LAST_5 = [
    61437.31013970,
    61409.77885185,
    61381.24752811,
    61352.71733871,
    61321.57890702,
];

const DEFAULT_PARAMS = {
    basePeriod: 113,
    volPeriod: 51,
    smoothing: true,
    smoothType: 3,
    smoothPeriod: 5
};

let testData;

test.before(async () => {
    testData = loadTestData();
});

test('VAMA accuracy', () => {

    const close = new Float64Array(testData.close);

    const result = wasm.vama_js(
        close,
        DEFAULT_PARAMS.basePeriod,
        DEFAULT_PARAMS.volPeriod,
        DEFAULT_PARAMS.smoothing,
        DEFAULT_PARAMS.smoothType,
        DEFAULT_PARAMS.smoothPeriod
    );

    assert.strictEqual(result.length, close.length, 'Output length should match input');


    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        EXPECTED_LAST_5,
        1e-6,
        "VAMA reference value mismatch"
    );
});

test('VAMA partial params', () => {

    const close = new Float64Array(testData.close);


    const result = wasm.vama_js(close, 113, 51, false, 1, 5);
    assert.strictEqual(result.length, close.length);
});

test('VAMA warmup NaN', () => {

    const close = new Float64Array(testData.close);

    const result = wasm.vama_js(
        close,
        DEFAULT_PARAMS.basePeriod,
        DEFAULT_PARAMS.volPeriod,
        DEFAULT_PARAMS.smoothing,
        DEFAULT_PARAMS.smoothType,
        DEFAULT_PARAMS.smoothPeriod
    );


    let firstValid = 0;
    for (let i = 0; i < close.length; i++) {
        if (!isNaN(close[i])) {
            firstValid = i;
            break;
        }
    }


    const warmup = firstValid + Math.max(DEFAULT_PARAMS.basePeriod, DEFAULT_PARAMS.volPeriod) - 1;


    if (warmup > 0) {
        const warmupValues = result.slice(0, Math.min(warmup, result.length));
        assertAllNaN(warmupValues, `Expected NaN in warmup period [0:${warmup})`);
    }


    const nonNaNCount = result.filter(v => !isNaN(v)).length;
    assert(nonNaNCount > 0, 'Should have some non-NaN values after warmup');
});

test('VAMA edge cases', () => {

    const data = new Float64Array([1.0, 2.0, 3.0, 4.0, 5.0]);

    const result = wasm.vama_js(data, 2, 2, false, 1, 5);

    assert.strictEqual(result.length, data.length, 'Output length should match input');




    const warmup = Math.max(2, 2) - 1;
    if (warmup > 0) {
        const warmupValues = result.slice(0, warmup);
        assertAllNaN(warmupValues, `Expected NaN in warmup [0:${warmup})`);
    }


    if (data.length > warmup) {
        const afterWarmup = result.slice(warmup);
        const hasValues = afterWarmup.some(v => !isNaN(v));
        assert(hasValues, 'Should have values after warmup');
    }
});

test('VAMA smoothing variations', () => {

    const close = new Float64Array(testData.close.slice(0, 100));

    const results = {};


    const smoothTypes = [
        { type: 1, name: 'SMA' },
        { type: 2, name: 'EMA' },
        { type: 3, name: 'WMA' }
    ];

    for (const { type, name } of smoothTypes) {
        const result = wasm.vama_js(close, 10, 5, true, type, 3);
        results[name] = result;
        assert.strictEqual(result.length, close.length, `${name} output length mismatch`);
    }


    const noSmooth = wasm.vama_js(close, 10, 5, false, 1, 5);


    for (const [name, smoothResult] of Object.entries(results)) {
        let foundDifference = false;
        for (let i = 0; i < smoothResult.length; i++) {
            if (!isNaN(smoothResult[i]) && !isNaN(noSmooth[i])) {
                if (Math.abs(smoothResult[i] - noSmooth[i]) > 1e-10) {
                    foundDifference = true;
                    break;
                }
            }
        }
        assert(foundDifference, `${name} smoothing should produce different values`);
    }
});

test('VAMA invalid periods', () => {

    const data = new Float64Array(10).fill(1.0);


    assert.throws(() => {
        wasm.vama_js(data, 0, 5, false, 1, 5);
    }, /Invalid period/, 'Should throw on zero base period');


    assert.throws(() => {
        wasm.vama_js(data, 5, 0, false, 1, 5);
    }, /Invalid period/, 'Should throw on zero vol period');


    assert.throws(() => {
        wasm.vama_js(data, 20, 5, false, 1, 5);
    }, /Invalid period/, 'Should throw when period exceeds data length');
});

test('VAMA invalid smooth type', () => {

    const data = new Float64Array(100).fill(1.0);


    assert.throws(() => {
        wasm.vama_js(data, 10, 5, true, 0, 5);
    }, /Invalid smooth/, 'Should throw on invalid smooth type 0');


    assert.throws(() => {
        wasm.vama_js(data, 10, 5, true, 4, 5);
    }, /Invalid smooth/, 'Should throw on invalid smooth type 4');
});

test('VAMA empty input', () => {

    const empty = new Float64Array([]);

    assert.throws(() => {
        wasm.vama_js(empty, 113, 51, false, 1, 5);
    }, /Empty|empty/, 'Should throw on empty array');
});

test('VAMA all NaN input', () => {

    const allNaN = new Float64Array(100).fill(NaN);

    assert.throws(() => {
        wasm.vama_js(allNaN, 113, 51, false, 1, 5);
    }, /All values are NaN|all NaN/, 'Should throw on all NaN values');
});

test('VAMA batch single params', () => {

    const close = new Float64Array(testData.close.slice(0, 200));

    const result = wasm.vama_batch(close, {
        base_period_range: [113, 113, 0],
        vol_period_range: [51, 51, 0]
    });

    assert(result.values, 'Should have values array');
    assert(result.combos, 'Should have combos array');
    assert.strictEqual(result.rows, 1, 'Should have 1 row');
    assert.strictEqual(result.cols, close.length, 'Cols should match data length');
    assert.strictEqual(result.combos.length, 1, 'Should have 1 combination');


    const batchRow = result.values.slice(0, close.length);
    const singleResult = wasm.vama_js(close, 113, 51, false, 3, 5);


    for (let i = 0; i < batchRow.length; i++) {
        if (!isNaN(batchRow[i]) && !isNaN(singleResult[i])) {
            assertClose(batchRow[i], singleResult[i], 1e-10,
                       `Batch vs single mismatch at ${i}`);
        }
    }
});

test('VAMA batch sweep', () => {

    const close = new Float64Array(testData.close.slice(0, 150));

    const result = wasm.vama_batch(close, {
        base_period_range: [100, 104, 2],
        vol_period_range: [40, 44, 2]
    });


    assert.strictEqual(result.rows, 9, 'Should have 9 rows');
    assert.strictEqual(result.cols, close.length, 'Cols should match data length');
    assert.strictEqual(result.combos.length, 9, 'Should have 9 combinations');


    const expectedBases = [100, 100, 100, 102, 102, 102, 104, 104, 104];
    const expectedVols = [40, 42, 44, 40, 42, 44, 40, 42, 44];

    for (let i = 0; i < result.combos.length; i++) {
        assert.strictEqual(result.combos[i].base_period, expectedBases[i],
                          `Base period mismatch at combo ${i}`);
        assert.strictEqual(result.combos[i].vol_period, expectedVols[i],
                          `Vol period mismatch at combo ${i}`);
    }
});

test('VAMA zero-copy functionality', () => {

    const close = new Float64Array(testData.close.slice(0, 200));
    const len = close.length;


    const inPtr = wasm.vama_alloc(len);
    const outPtr = wasm.vama_alloc(len);


    const memory = wasm.__wasm.memory;
    const inView = new Float64Array(memory.buffer, inPtr, len);
    const outView = new Float64Array(memory.buffer, outPtr, len);


    inView.set(close);


    wasm.vama_into(
        inPtr, outPtr, len,
        DEFAULT_PARAMS.basePeriod,
        DEFAULT_PARAMS.volPeriod,
        DEFAULT_PARAMS.smoothing,
        DEFAULT_PARAMS.smoothType,
        DEFAULT_PARAMS.smoothPeriod
    );


    const result = Array.from(outView);
    assert.strictEqual(result.length, close.length);


    const firstValidIdx = result.findIndex(v => !isNaN(v));
    assert(firstValidIdx > 0, 'Should have warmup NaN values');


    const regularResult = wasm.vama_js(
        close,
        DEFAULT_PARAMS.basePeriod,
        DEFAULT_PARAMS.volPeriod,
        DEFAULT_PARAMS.smoothing,
        DEFAULT_PARAMS.smoothType,
        DEFAULT_PARAMS.smoothPeriod
    );

    for (let i = 0; i < result.length; i++) {
        if (!isNaN(result[i]) && !isNaN(regularResult[i])) {
            assertClose(result[i], regularResult[i], 1e-10,
                       `Zero-copy mismatch at index ${i}`);
        }
    }


    wasm.vama_free(inPtr, len);
    wasm.vama_free(outPtr, len);
});

test('VAMA with real market data', () => {

    const candles = loadTestData();
    const close = new Float64Array(candles.close);

    const result = wasm.vama_js(
        close,
        DEFAULT_PARAMS.basePeriod,
        DEFAULT_PARAMS.volPeriod,
        DEFAULT_PARAMS.smoothing,
        DEFAULT_PARAMS.smoothType,
        DEFAULT_PARAMS.smoothPeriod
    );

    assert.strictEqual(result.length, close.length);


    const validValues = result.filter(v => !isNaN(v));
    assert(validValues.length > 0, 'Should have valid values');


    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        EXPECTED_LAST_5,
        1e-6,
        "Real data should produce expected reference values"
    );
});

test('VAMA performance characteristics', () => {

    const sizes = [100, 200, 400, 800];
    const times = [];

    for (const size of sizes) {

        const data = new Float64Array(testData.close.slice(0, size));

        const start = performance.now();
        wasm.vama_js(data, 20, 10, false, 1, 5);
        const end = performance.now();

        times.push(end - start);
    }


    for (let i = 1; i < times.length; i++) {
        const sizeRatio = sizes[i] / sizes[i-1];
        const timeRatio = times[i] / times[i-1];

        assert(timeRatio < sizeRatio * 3,
               `Performance should scale linearly (${timeRatio} vs ${sizeRatio})`);
    }
});

File: tests/python/test_volatility_adjusted_ma.py
"""
Python binding tests for VAMA (Volatility Adjusted Moving Average) indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:
    pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close


class TestVama:
    """Test suite for VAMA indicator Python bindings"""
    
    
    EXPECTED_LAST_5 = [
        61437.31013970,
        61409.77885185,
        61381.24752811,
        61352.71733871,
        61321.57890702,
    ]
    
    DEFAULT_PARAMS = {
        'base_period': 113,
        'vol_period': 51,
        'smoothing': True,
        'smooth_type': 3,  
        'smooth_period': 5
    }
    
    @pytest.fixture(scope='class')
    def test_data(self):
        """Load test data once for all tests"""
        return load_test_data()
    
    def test_vama_accuracy(self, test_data):
        """Test VAMA matches expected reference values - mirrors test_vama_accuracy"""
        close = test_data['close']
        
        result = ta_indicators.vama(
            close,
            **self.DEFAULT_PARAMS
        )
        
        assert len(result) == len(close), "Output length should match input"
        
        
        last_5 = result[-5:]
        for i, (actual, expected) in enumerate(zip(last_5, self.EXPECTED_LAST_5)):
            if not np.isnan(actual):
                assert_close(
                    actual, 
                    expected, 
                    rtol=1e-6,  
                    msg=f"VAMA reference value mismatch at index {i}"
                )
    
    def test_vama_partial_params(self, test_data):
        """Test VAMA with partial parameters - mirrors check_vama_partial_params"""
        close = test_data['close']
        
        
        result = ta_indicators.vama(
            close,
            base_period=113,
            vol_period=51
        )
        assert len(result) == len(close)
    
    def test_vama_warmup_nan(self, test_data):
        """Test VAMA warmup period NaN handling - mirrors check_vama_warmup_nan"""
        close = test_data['close']
        
        result = ta_indicators.vama(
            close,
            **self.DEFAULT_PARAMS
        )
        
        
        first_valid = next(i for i, v in enumerate(close) if not np.isnan(v))
        
        
        warmup = first_valid + max(self.DEFAULT_PARAMS['base_period'], 
                                  self.DEFAULT_PARAMS['vol_period']) - 1
        
        
        if warmup > 0:
            warmup_values = result[:min(warmup, len(result))]
            assert np.all(np.isnan(warmup_values)), f"Expected NaN in warmup period [0:{warmup})"
        
        
        non_nan_count = np.count_nonzero(~np.isnan(result))
        assert non_nan_count > 0, "Should have some non-NaN values after warmup"
    
    def test_vama_edge_cases(self):
        """Test VAMA edge cases - mirrors check_vama_edge_cases"""
        
        data = np.array([1.0, 2.0, 3.0, 4.0, 5.0])
        
        result = ta_indicators.vama(
            data,
            base_period=2,
            vol_period=2,
            smoothing=False
        )
        
        assert len(result) == len(data), "Output length should match input"
        
        
        
        
        warmup = max(2, 2) - 1  
        if warmup > 0:
            assert np.all(np.isnan(result[:warmup])), f"Expected NaN in warmup [0:{warmup})"
        
        
        if len(data) > warmup:
            assert not np.all(np.isnan(result[warmup:])), "Should have values after warmup"
    
    def test_vama_smoothing_variations(self, test_data):
        """Test different smoothing types - mirrors check_vama_smoothing"""
        close = test_data['close'][:100]  
        
        results = {}
        
        
        for smooth_type, name in [(1, 'SMA'), (2, 'EMA'), (3, 'WMA')]:
            result = ta_indicators.vama(
                close,
                base_period=10,
                vol_period=5,
                smoothing=True,
                smooth_type=smooth_type,
                smooth_period=3
            )
            results[name] = result
            assert len(result) == len(close), f"{name} output length mismatch"
        
        
        no_smooth = ta_indicators.vama(
            close,
            base_period=10,
            vol_period=5,
            smoothing=False
        )
        
        
        for name, smooth_result in results.items():
            valid_idx = ~(np.isnan(smooth_result) | np.isnan(no_smooth))
            if np.any(valid_idx):
                assert not np.allclose(smooth_result[valid_idx], no_smooth[valid_idx], rtol=1e-10), \
                    f"{name} smoothing should produce different values"
    
    def test_vama_invalid_periods(self):
        """Test VAMA fails with invalid periods - mirrors check_vama_period_errors"""
        data = np.array([1.0] * 10)
        
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.vama(data, base_period=0, vol_period=5)
        
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.vama(data, base_period=5, vol_period=0)
        
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.vama(data, base_period=20, vol_period=5)
    
    def test_vama_invalid_smooth_type(self):
        """Test VAMA fails with invalid smooth type"""
        data = np.ones(100)
        
        
        with pytest.raises(ValueError, match="Invalid smooth"):
            ta_indicators.vama(
                data, 
                base_period=10, 
                vol_period=5,
                smoothing=True,
                smooth_type=0  
            )
        
        
        with pytest.raises(ValueError, match="Invalid smooth"):
            ta_indicators.vama(
                data,
                base_period=10,
                vol_period=5,
                smoothing=True,
                smooth_type=4  
            )
    
    def test_vama_empty_input(self):
        """Test VAMA fails with empty input - mirrors test_vama_empty_input"""
        empty = np.array([])
        
        with pytest.raises(ValueError, match="Empty|empty"):
            ta_indicators.vama(empty)
    
    def test_vama_all_nan_input(self):
        """Test VAMA fails with all NaN input - mirrors test_vama_all_nan"""
        all_nan = np.full(100, np.nan)
        
        with pytest.raises(ValueError, match="All values are NaN|all NaN"):
            ta_indicators.vama(all_nan)
    
    def test_vama_streaming(self, test_data):
        """Test VAMA streaming matches batch calculation - mirrors test_vama_stream"""
        close = test_data['close'][:200]  
        
        
        batch_result = ta_indicators.vama(
            close,
            **self.DEFAULT_PARAMS
        )
        
        
        stream = ta_indicators.VamaStream(
            **self.DEFAULT_PARAMS
        )
        
        stream_values = []
        for price in close:
            result = stream.update(price)
            stream_values.append(result if result is not None else np.nan)
        
        stream_values = np.array(stream_values)
        
        
        assert len(batch_result) == len(stream_values)
        
        
        warmup = max(self.DEFAULT_PARAMS['base_period'], 
                    self.DEFAULT_PARAMS['vol_period'])
        
        for i in range(warmup, len(batch_result)):
            if not np.isnan(batch_result[i]) and not np.isnan(stream_values[i]):
                assert_close(
                    batch_result[i], 
                    stream_values[i], 
                    rtol=2e-2,  
                    msg=f"Streaming mismatch at index {i}"
                )
    
    def test_vama_batch_single_params(self, test_data):
        """Test VAMA batch with single parameter set - mirrors check_batch_default_row"""
        close = test_data['close']
        
        result = ta_indicators.vama_batch(
            close,
            base_period_range=(113, 113, 0),
            vol_period_range=(51, 51, 0)
        )
        
        assert 'values' in result
        assert 'base_periods' in result
        assert 'vol_periods' in result
        
        
        assert len(result['base_periods']) == 1
        assert len(result['vol_periods']) == 1
        assert result['values'].shape == (1, len(close))
        
        
        
        batch_row = result['values'][0]
        single_result = ta_indicators.vama(
            close,
            base_period=113,
            vol_period=51,
            smoothing=False  
        )
        
        
        for i, (b, s) in enumerate(zip(batch_row, single_result)):
            if not np.isnan(b) and not np.isnan(s):
                assert_close(b, s, rtol=1e-10, 
                           msg=f"Batch vs single mismatch at {i}")
    
    def test_vama_batch_sweep(self, test_data):
        """Test VAMA batch parameter sweep - mirrors check_batch_sweep"""
        close = test_data['close'][:100]  
        
        result = ta_indicators.vama_batch(
            close,
            base_period_range=(40, 44, 2),  
            vol_period_range=(20, 24, 2)      
        )
        
        
        assert len(result['base_periods']) == 9
        assert len(result['vol_periods']) == 9
        assert result['values'].shape == (9, len(close))
        
        
        expected_bases = [40, 40, 40, 42, 42, 42, 44, 44, 44]
        expected_vols = [20, 22, 24] * 3
        
        for i in range(9):
            assert result['base_periods'][i] == expected_bases[i]
            assert result['vol_periods'][i] == expected_vols[i]
    
    def test_vama_kernel_consistency(self, test_data):
        """Test different kernels produce consistent results"""
        close = test_data['close'][:100]  
        
        kernels = ['scalar', 'sse2']  
        results = {}
        
        for kernel in kernels:
            try:
                result = ta_indicators.vama(
                    close,
                    base_period=20,
                    vol_period=10,
                    kernel=kernel
                )
                results[kernel] = result
            except ValueError:
                
                pass
        
        
        if len(results) > 1:
            kernel_names = list(results.keys())
            for i in range(1, len(kernel_names)):
                valid_idx = ~(np.isnan(results[kernel_names[0]]) | 
                            np.isnan(results[kernel_names[i]]))
                if np.any(valid_idx):
                    np.testing.assert_allclose(
                        results[kernel_names[0]][valid_idx],
                        results[kernel_names[i]][valid_idx],
                        rtol=1e-10,
                        err_msg=f"Kernel {kernel_names[0]} vs {kernel_names[i]} mismatch"
                    )


if __name__ == '__main__':
    pytest.main([__file__, '-v'])